### ISSUE TITLE
Error when connecting to a service in a namespace that has a headless service

### ISSUE THREAD

**Issue by justinrush** [microsoft/mindaro#155]
**Describe the bug**
I've seen this a few times now and just wanted to write it up and see if there is a workaround or a fix in the works.  When you connect Bridge up to a service that is in the same namespace as a headless service (for example, for an sts), you get an error complaining about a duplicate key, presumably due to how the headless service is handled (see logs below).
 
**To Reproduce**
Steps to follow to reproduce this issue.
Create a simple deployment and service in a namespace that contains an existing set of services. Common ones I see for sts deploys is:
* a clusterIP service called myservice
* an external LB service called myservice-external
* a headless service called myservice-headless

Attempt to bridge to your simple service and find that an error is generated.

 
**Expected behavior**
I expect the bridge to be created normally and not crash
 
**Logs**
the first service is the one I'm bridging to, the others are services that exist in the same namespace.
```
2021-04-08T21:18:26.1896771Z | Library | TRACE | ServicePortForwarder started on dev-service:8080 on local port 8080
2021-04-08T21:18:26.1901583Z | Library | TRACE | ServicePortForwarder started on liftbridge:9292 on local port 9292
2021-04-08T21:18:26.1905459Z | Library | TRACE | ServicePortForwarder started on liftbridge-external:9292 on local port 9292
2021-04-08T21:18:26.1908088Z | Library | TRACE | ServicePortForwarder started on liftbridge:9292 on local port 9292
2021-04-08T21:18:26.1910385Z | Library | TRACE | Event: LocalEnvironmentManager-StartServiceRouters <json>{"properties":{"result":"Failed"},"metrics":{"durationInMs":2290.0}}</json>
2021-04-08T21:18:26.1918480Z | Library | TRACE | Event: ConnectManagementClient-ConfigureLocalHost <json>{"properties":{"containerName":"ef260e9aa3c6","numOfReachableServices":"9","numOfEnvironmentVariables":"8","numOfContainerPorts":"1","numOfVolumeMounts":"1","result":"Failed"},"metrics":{"durationInMs":2296.0}}</json>
2021-04-08T21:18:26.1922100Z | Library | ERROR | Failed to configure local host.
2021-04-08T21:18:26.2520837Z | Library | ERROR | Logging handled exception: System.ArgumentException: {"ClassName":"System.ArgumentException","Message":"An item with the same key has already been added. Key: liftbridge","Data":null,"InnerException":null,"HelpURL":null,"StackTraceString":"   at System.Collections.Generic.Dictionary`2.TryInsert(TKey key, TValue value, InsertionBehavior behavior)\r\n   at Microsoft.BridgeToKubernetes.Library.Connect.LocalEnvironmentManager.StartServiceRoutersInnerAsync(String devhostAgentClientTarget, ContainerWorkloadInfo workloadInfo, IEnumerable`1 elevationRequests, IProgress`1 progress, CancellationToken cancellationToken)\r\n   at Microsoft.BridgeToKubernetes.Library.Connect.LocalEnvironmentManager.StartServiceRoutersAsync(String devhostAgentClientTarget, ContainerWorkloadInfo workloadInfo, IEnumerable`1 elevationRequests, IProgress`1 progress, CancellationToken cancellationToken)\r\n   at Microsoft.BridgeToKubernetes.Library.ManagementClients.ConnectManagementClient.<>c__DisplayClass16_0.<<ConfigureLocalHostAsync>b__0>d.MoveNext()\r\n--- End of stack trace from previous location where exception was thrown ---\r\n   at Microsoft.BridgeToKubernetes.Library.ManagementClients.ManagementClientExceptionStrategy.RunWithHandlingAsync[T](Func`1 func, FailureConfig failureConfig)","RemoteStackTraceString":null,"RemoteStackIndex":0,"ExceptionMethod":null,"HResult":-2147024809,"Source":"System.Private.CoreLib","WatsonBuckets":null,"ParamName":null}
```
 
**Environment Details**
Client used (VS Code/Visual Studio): VS Code
 mindaro.mindaro@1.0.120210401
Client's version:
Version: 1.55.0 (user setup)
Commit: c185983a683d14c396952dd432459097bc7f757f
Date: 2021-03-30T16:01:55.261Z
Electron: 11.3.0
Chrome: 87.0.4280.141
Node.js: 12.18.3
V8: 8.7.220.31-electron.0
OS: Windows_NT x64 10.0.19042

**Additional context**
Add any other outputs from the clients or context you would like to share.

Depending on what you are trying to do, a workaround could be to just deploy the service in another namespace.

**Comment by daniv-msft** [comment]
Adding @amsoedal who is looking at customer issues this week, and @lolodi who I expect has context on that.
This looks like the issue you fixed recently @lolodi, right? If that's the case, then it should be fixed with our next release, on April 19th.

**Comment by lolodi** [comment]
@daniv-msft this is correct, this is the issue I fixed recently and that will be part of our next release.

**Comment by justinrush** [comment]
OK, thanks, i'll test the new release when it comes out.

**Comment by sidgujrathi** [comment]
I am also facing the same issue. I will test once the new release comes out! Any dates for upcoming realses?

**Comment by irperez** [comment]
Hey guys this is an issue for us, but the work around (service in a separate namespace) is not so simple for us as we have a lot of developers.  One dev per namespace.  Any ideas?  When can we expect the next release?

**Comment by irperez** [comment]
For the sake of clarity and documentation, we're seeing this with RabbitMQ.

> 2021-04-22T15:40:09.1545889Z | Library | TRACE | ServicePortForwarder started on rabbitmq:5672 on local port 5672
2021-04-22T15:40:09.1548490Z | Library | TRACE | ServicePortForwarder started on rabbitmq:4369 on local port 4369
2021-04-22T15:40:09.1552267Z | Library | TRACE | ServicePortForwarder started on rabbitmq:25672 on local port 25672
2021-04-22T15:40:09.1555449Z | Library | TRACE | ServicePortForwarder started on rabbitmq:15672 on local port 15672
2021-04-22T15:40:09.1558213Z | Library | TRACE | ServicePortForwarder started on security:80 on local port 80
2021-04-22T15:40:09.1560077Z | Library | TRACE | ServicePortForwarder started on rabbitmq:5672 on local port 5672
2021-04-22T15:40:09.1562518Z | Library | TRACE | ServicePortForwarder started on rabbitmq:4369 on local port 4369
2021-04-22T15:40:09.1564991Z | Library | TRACE | ServicePortForwarder started on rabbitmq:25672 on local port 25672
2021-04-22T15:40:09.1567558Z | Library | TRACE | ServicePortForwarder started on rabbitmq:15672 on local port 15672
2021-04-22T15:40:09.1570629Z | Library | TRACE | Event: LocalEnvironmentManager-StartServiceRouters <json>{"properties":{"result":"Failed"},"metrics":{"durationInMs":2192.0}}</json>
2021-04-22T15:40:09.1582463Z | Library | TRACE | Event: ConnectManagementClient-ConfigureLocalHost <json>{"properties":{"containerName":"d12a9b6357a1","numOfReachableServices":"15","numOfEnvironmentVariables":"53","numOfContainerPorts":"1","numOfVolumeMounts":"2","result":"Failed"},"metrics":{"durationInMs":2199.0}}</json>
2021-04-22T15:40:09.1587814Z | Library | ERROR | Failed to configure local host.
2021-04-22T15:40:09.1796679Z | Library | ERROR | Logging handled exception: System.ArgumentException: {"ClassName":"System.ArgumentException","Message":"An item with the same key has already been added. Key: rabbitmq","Data":null,"InnerException":null,"HelpURL":null,"StackTraceString":"   at System.Collections.Generic.Dictionary`2.TryInsert(TKey key, TValue value, InsertionBehavior behavior)\r\n   at Microsoft.BridgeToKubernetes.Library.Connect.LocalEnvironmentManager.StartServiceRoutersInnerAsync(String devhostAgentClientTarget, ContainerWorkloadInfo workloadInfo, IEnumerable`1 elevationRequests, IProgress`1 progress, CancellationToken cancellationToken)\r\n   at Microsoft.BridgeToKubernetes.Library.Connect.LocalEnvironmentManager.StartServiceRoutersAsync(String devhostAgentClientTarget, ContainerWorkloadInfo workloadInfo, IEnumerable`1 elevationRequests, IProgress`1 progress, CancellationToken cancellationToken)\r\n   at Microsoft.BridgeToKubernetes.Library.ManagementClients.ConnectManagementClient.<>c__DisplayClass16_0.<<ConfigureLocalHostAsync>b__0>d.MoveNext()\r\n--- End of stack trace from previous location where exception was thrown ---\r\n   at Microsoft.BridgeToKubernetes.Library.ManagementClients.ManagementClientExceptionStrategy.RunWithHandlingAsync[T](Func`1 func, FailureConfig failureConfig)","RemoteStackTraceString":null,"RemoteStackIndex":0,"ExceptionMethod":null,"HResult":-2147024809,"Source":"System.Private.CoreLib","WatsonBuckets":null,"ParamName":null}

**Comment by lolodi** [comment]
Hi @irperez!
We are targeting next week for the official prod release, in the mean time you can get unblocked by using our staging bits.
To do that you can start VSCode with BRIDGE_ENVIRONMENT=staging environment variable set. Let us know if it works for you.

**Comment by amsoedal** [comment]
Hi @irperez @justinrush @sidgujrathi, 
Before you try @lolodi's fix in the previous comment, could you please delete this folder:

On Windows: %UserProfile%\AppData\Roaming\Code\User\globalStorage\mindaro.mindaro\file-downloader-downloads
On MacOS: "/Users/your_username/Library/Application Support/Code/User/globalStorage/mindaro.mindaro/file-downloader-downloads"
On Linux: /home/your_username/.config/Code/User/globalStorage/mindaro.mindaro/file-downloader-downloads

**Comment by irperez** [comment]
@lolodi we're using Visual Studio.

**Comment by irperez** [comment]
@amsoedal We tried deleting that folder, but we don't have the same paths.

![MicrosoftTeams-image (3)](https://user-images.githubusercontent.com/1376080/115767459-2dd06580-a377-11eb-9fc3-f72ec2eb382f.png)

**Comment by lolodi** [comment]
@irperez Yeah those paths are for VSCode, if you are using VS you'll need to wait the next VS release, probably in the next week or two.

**Comment by irperez** [comment]
@lolodi is there a way to download a previous version of the plugin?  I don't see a way on the marketplace for visual studio.  This seems to solve our problem.  As others that are using an older version don't see the issue.

**Comment by lolodi** [comment]
@irperez 
You can download the previous version here:
https://marketplace.visualstudio.com/_apis/public/gallery/publishers/ms-azuretools/vsextensions/mindaro/2.1.20210327.1/vspackage

Let us know if this works for you

**Comment by irperez** [comment]
@lolodi  I have this version installed (I never upgraded) and its fine for me.  However other team members are saying that once they've upgraded, they try to download and install, it still doesn't work.  Can you confirm this link isn't installing the latest and greatest?  Is there a way to version check it?

**Comment by irperez** [comment]
@lolodi I just confirmed with my team, using the link you provided installs the latest version.  👎🏽

**Comment by justinrush** [comment]
> @lolodi I just confirmed with my team, using the link you provided installs the latest version. 👎🏽

you can just go to the version history tab on [this site](https://marketplace.visualstudio.com/items?itemName=mindaro.mindaro), download the newest one, and install as an offline extension

**Comment by lolodi** [comment]
@justinrush 
Unfortunately that one is for VSCode not VS so it doesn't work for @irperez 
@irperez ,
You are correct, was our understanding that modifying the url with different versions would provide different versions. We are fetching the 327 release from our releases and I'm going to drop a link in a few minutes.

**Comment by lolodi** [comment]
@irperez ,
I uploaded a copy of our previous Visusal Studio extension here: https://bridgetokubernetes.blob.core.windows.net/vs-releases/Microsoft.VisualStudio.Kubernetes.Debugging.Vsix.2.1.20210327.1.vsix

This should work.
Let me know if you have any problem.

**Comment by irperez** [comment]
@lolodi That did the trick!!!  Thank you so much for your responsiveness.  

We can now easily wait for the next update for VStudio.  

Thanks for the feature itself and keep up the great work!  Our team over here really depends on it and its gaining a wider audience with our product engineering.  Everyone that sees it in action is impressed!

**Comment by lolodi** [comment]
Great!!
Glad to know it worked, and even more happy to know you and your team like our product!
Keep spreading the word :)
